[18.0][MIG] l10n_es_vat_prorate: Migration to 18.0

BetoRaja opened the pull request:
We migrated the VAT prorate module from Odoo 17 to 18, adapting the old _compute_all_tax logic to the new flow using create, write, and button_draft.
We generate prorate journal lines with display_type='tax' to keep them out of the invoice lines and avoid UI clutter.

@HaraldPanten @Jaimermaccione @ValentinVinagre 
[T-8592]

etobella committed:
l10n_es_vat_prorate

sanchonuria committed:
l10n_es_vat_prorate: black, isort, prettier

sanchonuria committed:
l10n_es_vat_prorate: Migration to 14.0

etobella committed:
l10n_es_vat_prorrate: Allow to manage directly with l10n_es_aeat_sii_oca

etobella committed:
l10n_es_vat_prorate: signature was wrong

pedrobaeza committed:
l10n_es_aeat*: VAT prorate proper encapsulation + names + tests

pedrobaeza committed:
l10n_es_vat_prorate: Handle analytic information

The VAT prorate tax line should contain the analytic info from the
expense line.

With this commit, we put such information, and make sure that is
synchronized on any change.

TT41832

etobella committed:
l10n_es_vat_prorate: Onchange should apply on invoice line, not lines

pedrobaeza committed:
l10n_es_vat_prorate: Adapt tests to the revert of the inclusion of s_iva0_ns

etobella committed:
l10n_es_vat_prorate: Manage intracommunity and extracommunity taxes

pedrobaeza committed:
l10n_es_vat_prorate: Migration to 15.0

pedrobaeza committed:
l10n_es_vat_prorate: Handle properly prorate refund account

When doing a refund, the account of the prorate line is changed to the
tax repartition line default one (472).

That's because there's some code on the method `_reverse_move_vals`
that is checking `tax_repartition_line_id` value in the returned
dictionary for ignoring the copied value. See
https://github.com/odoo/odoo/blob/82b1e68b60eb14896666adfd3ad1f753e944d393/addons/account/models/account_move.py#L2515-L2533

What we do for tricking it (as it's atomic and can't be intercepted in
other part) is:

- When copying values, we move the `tax_repartition_line_id` into
  another field name.
- Thus, `_reverse_move_vals` is not modifying the account.
- Immediately after returning from `_reverse_move_vals`, we restored
  the field value.

TT44656

pedrobaeza committed:
l10n_es_vat_prorate: Migration to 16.0

The dynamic lines management has totally changed in this version, so the
code has been rewritten mostly from scratch, but now is cleaner thanks
to the avoid of a lot of hooks.

Some JSON from tests were not correct in previous version due to
rounding, and now they are OK: 10 / 1.2 = 8.3333333, which is 8.33, not
8.34.

TT43236

Commit:
Update l10n_es_vat_prorate.pot

OCA-git-bot committed:
README.rst

OCA-git-bot committed:
README.rst

carolinafernandez-tecnativa committed:
l10n_es_vat_prorate: Restrict vat prorate value

Commit:
Update l10n_es_vat_prorate.pot

OCA-git-bot committed:
README.rst

OCA-git-bot committed:
l10n_es_vat_prorate 16.0.1.0.1

OCA-git-bot committed:
README.rst

Ivorra78 committed:
Added translation using Weblate (Spanish)

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (24 of 24 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_vat_prorate/es/

BrunoBailo committed:
Added translation using Weblate (Catalan)

BrunoBailo committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (24 of 24 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_vat_prorate/ca/

pedrobaeza committed:
l10n_es_vat_prorate: Indicate incompatibility with cash basis

OCA-git-bot committed:
README.rst

manuelregidor committed:
l10n_es_vat_prorate: black, isort, prettier

manuelregidor committed:
l10n_es_vat_prorate: Migration to 17.0

Commit:
Update l10n_es_vat_prorate.pot

OCA-git-bot committed:
post-merge updates

Tisho99 committed:
l10n_es_vat_prorate: add special vat prorrate

Commit:
Update l10n_es_vat_prorate.pot

OCA-git-bot committed:
post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-17.0/l10n-spain-17.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-17-0/l10n-spain-17-0-l10n_es_vat_prorate/

Tisho99 committed:
l10n_es_vat_prorate: move prorate type to prorate level and add constraint

Tisho99 committed:
l10n_es_aeat_mod303_vat_prorate: adaptation to dependences refactor

Commit:
Update l10n_es_vat_prorate.pot

OCA-git-bot committed:
post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: l10n-spain-17.0/l10n-spain-17.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-17-0/l10n-spain-17-0-l10n_es_vat_prorate/

Tisho99 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (36 of 36 strings)

Translation: l10n-spain-17.0/l10n-spain-17.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-17-0/l10n-spain-17-0-l10n_es_vat_prorate/es/

Tisho99 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (36 of 36 strings)

Translation: l10n-spain-17.0/l10n-spain-17.0-l10n_es_vat_prorate
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-17-0/l10n-spain-17-0-l10n_es_vat_prorate/ca/

pedrobaeza committed:
l10n_es_*: Fix intra/extra-community taxes refs and new exempt taxes

Odoo screwed up the tax XML-IDs in the middle of the 17.0 lifecycle,
introducing an unfolding of the exempt taxes, using existing ones
that belongs to the sales intra and extra community taxes.

After that, they have renamed the XML-IDs of both taxes to new ones.

We need to handle this in 2 ways:

- Detect and correct existing installations previous to this
  taxpocalypse for renaming them.
- Change the reference in all places to the new XML-IDs.

Besides, we have to add the unfolded exempt taxes in the reports.

OCA-git-bot committed:
post-merge updates

Jaimermaccione reviewed (changes requested):
@BetoRaja I've been running some tests with the module, and it doesn't seem to be working as expected.

---

#### Configuration tested

- Enabled **"With VAT prorate"** at the **company** level.
- Created several **VAT prorate records** (both *general* and *special*).
- Marked the **"With VAT prorate"** option on the **tax** used in the invoices.

---

####  Test scenarios

- Created **purchase orders** and converted them into **vendor bills**.
- Created **manual vendor bills** (not linked to POs).
- In every case, the **VAT is applied in full** — the prorated amount is not calculated.

---

####  Extra observations

- I added the `prorate_id` field to the invoice form:
  - It correctly detects a prorate from the company setup.

 ![image](https://github.com/user-attachments/assets/c1274855-da31-499f-a862-e6a9ced69644)

- Also made the **"Is VAT prorate?"** field (`vat_prorate`) visible in the journal items:
  - No line has it activated.

  ![image](https://github.com/user-attachments/assets/cd65512f-1f7c-4646-8cb2-a04581aac0ef)

---

####  Version comparison

- @Jesarregui tested this in **Odoo 17**, where prorate works correctly.
- In **Odoo 18**, the same setup does not produce the expected result.

---

####  Quick look at GitHub

Trying to understand the difference, I noticed that the method `_compute_all_tax` in `account.move` — which appears in the module’s code — is no longer present in Odoo 18. That might be related to the issue.

Here’s the part of the module where it shows up:
[OCA/l10n-spain PR #4251 — account_move.py (lines 67–74)](https://github.com/OCA/l10n-spain/pull/4251/files#diff-b98482b0eba53728bc7ae07431251e2bd1d61f5fc7001e15a80316693efe1b74R67-R74)

And due to the changes made that match what I see in v18 locally, this might be the commit responsible where the method was altered to improve the global rounding calculation:  
[Odoo commit: 73d585a85b093db07e4eb5541aaa03e74af5fd62 (lines 911–964)](https://github.com/odoo/odoo/commit/73d585a85b093db07e4eb5541aaa03e74af5fd62#diff-78f3e1847de8ca0acf28d72a412947a549acdb08142d1dadf7646363d7972cb0L911-L964)

Hope this helps identify what's going on!

HaraldPanten commented:
/ocabot migration l10n_es_vat_prorate

BetoRaja commented:
> @BetoRaja I've been running some tests with the module, and it doesn't seem to be working as expected.
> 
> #### Configuration tested
> * Enabled **"With VAT prorate"** at the **company** level.
> * Created several **VAT prorate records** (both _general_ and _special_).
> * Marked the **"With VAT prorate"** option on the **tax** used in the invoices.
> 
> #### Test scenarios
> * Created **purchase orders** and converted them into **vendor bills**.
> * Created **manual vendor bills** (not linked to POs).
> * In every case, the **VAT is applied in full** — the prorated amount is not calculated.
> 
> #### Extra observations
> * I added the `prorate_id` field to the invoice form:
>   
>   * It correctly detects a prorate from the company setup.
> 
> ![image](https://private-user-images.githubusercontent.com/153611478/462318674-c1274855-da31-499f-a862-e6a9ced69644.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIwNTA0MDEsIm5iZiI6MTc1MjA1MDEwMSwicGF0aCI6Ii8xNTM2MTE0NzgvNDYyMzE4Njc0LWMxMjc0ODU1LWRhMzEtNDk5Zi1hODYyLWU2YTljZWQ2OTY0NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQwODM1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YjUwYmRjODY2ODRlZDk3Mjc0YmM1ZDY5YmNlYzYyMjYxMTBkMmU0NTRlMzA1Y2Y1YTZkMWUwMTQyYWYyMDg5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.yWt34FqUnFZUtnmWrwK9XZi-i-ZQzMciqRx_LXaF9VE)
> 
> * Also made the **"Is VAT prorate?"** field (`vat_prorate`) visible in the journal items:
>   
>   * No line has it activated.
>   
>   ![image](https://private-user-images.githubusercontent.com/153611478/462192903-cd65512f-1f7c-4646-8cb2-a04581aac0ef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIwNTA0MDEsIm5iZiI6MTc1MjA1MDEwMSwicGF0aCI6Ii8xNTM2MTE0NzgvNDYyMTkyOTAzLWNkNjU1MTJmLTFmN2MtNDY0Ni04Y2IyLWEwNDU4MWFhYzBlZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQwODM1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMzA3YzFhOWM4ZTgwMTA5ZmQzYjY4MzljMDM0ZjY2NjM0NTlhNDljMjZlNjkzYzU1M2NiYjBmZmI2ZjFmMWRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3ejqVAoH4_lD5u4867O5p-ssHJzrmTpAYXvOGYh-3KQ)
> 
> #### Version comparison
> * @Jesarregui tested this in **Odoo 17**, where prorate works correctly.
> * In **Odoo 18**, the same setup does not produce the expected result.
> 
> #### Quick look at GitHub
> Trying to understand the difference, I noticed that the method `_compute_all_tax` in `account.move` — which appears in the module’s code — is no longer present in Odoo 18. That might be related to the issue.
> 
> Here’s the part of the module where it shows up: [OCA/l10n-spain PR #4251 — account_move.py (lines 67–74)](https://github.com/OCA/l10n-spain/pull/4251/files#diff-b98482b0eba53728bc7ae07431251e2bd1d61f5fc7001e15a80316693efe1b74R67-R74)
> 
> And due to the changes made that match what I see in v18 locally, this might be the commit responsible where the method was altered to improve the global rounding calculation: [Odoo commit: 73d585a85b093db07e4eb5541aaa03e74af5fd62 (lines 911–964)](https://github.com/odoo/odoo/commit/73d585a85b093db07e4eb5541aaa03e74af5fd62#diff-78f3e1847de8ca0acf28d72a412947a549acdb08142d1dadf7646363d7972cb0L911-L964)
> 
> Hope this helps identify what's going on!

@Jaimermaccione Changes have been made to make it work as it should in v18.
Changes made to support VAT prorate in Odoo 18:

- Removed `_compute_all_tax` logic, since it's no longer used in Odoo 18.
- Moved the prorate logic to `action_post()` so it's applied when the invoice is validated.
- Adjusted tax lines (`tax_line_id`) based on the active prorate percentage.
- Created an additional journal line for the non-deductible portion (`vat_prorate = True`).
- Avoided modifying `tax_ids`, preserving correct tax reporting and reuse.
- Works for both general and special prorate types.

 Tested with:
- Vendor bills (manual and from POs).
- Different prorate settings at company level.

Jaimermaccione commented:
@BetoRaja I've been reviewing the module with the changes you've introduced, and I'm trying to get the VAT prorata to compute, but it's not being calculated.

I'm using the same context as in the previous review:

- Multiple prorata lines are active per company.
- The company has **prorata enabled**.
- The tax has the **"With VAT prorata"** checkbox selected.
- On the **vendor bill**, a prorata is clearly set.

Despite this setup, the prorata amount is not being computed.

Jaimermaccione reviewed (changes requested):
@BetoRaja Reviewed, thanks! Right now it seems to be working well in terms of the calculations.

- VAT prorate is calculated **correctly** when **validating supplier invoices**.
- Deductible and non-deductible VAT amounts are accurately split.

But I see there are some changes compared to previous versions:
- **Before (≤ Odoo 17):** Prorate applied in **draft** and updated live on edits.
- **Now (Odoo 18):** Prorate applies **only on validation**.

This could have an impact on:
- Users may expect prorate to show earlier (as before).
- Risk of confusion or oversight before confirming invoices.

I suggest keeping this behavior available during the draft stage, in case users need to view and work with this information before validating the invoice.

pedrobaeza committed:
[FIX] l10n_es_vat_prorate: Avoid false prorate lines

When you publish a vendor bill with prorate taxes, and then reset it to
draft and change amounts, there's a chance that Odoo rewrites the
move lines switching the prorated expense and the tax, causing that
the new tax line, being the old expense one, have the flag `vat_prorate`
set, and thus not being included in AEAT reports.

As it seems we can't do anything for assuring the same move lines
processing order upstream, what we do is to force in all the entries
the False value, and then only putting to True when building the
VAT prorate expense line.

TT57089

pedrobaeza commented:
Mirad si #4279 sigue siendo relevante para 18

BetoRaja commented:
> @BetoRaja Reviewed, thanks! Right now it seems to be working well in terms of the calculations.
> 
> * VAT prorate is calculated **correctly** when **validating supplier invoices**.
> * Deductible and non-deductible VAT amounts are accurately split.
> 
> But I see there are some changes compared to previous versions:
> 
> * **Before (≤ Odoo 17):** Prorate applied in **draft** and updated live on edits.
> * **Now (Odoo 18):** Prorate applies **only on validation**.
> 
> This could have an impact on:
> 
> * Users may expect prorate to show earlier (as before).
> * Risk of confusion or oversight before confirming invoices.
> 
> I suggest keeping this behavior available during the draft stage, in case users need to view and work with this information before validating the invoice.

@Jaimermaccione Thanks for your feedback!

The requested changes have been implemented.

Prorate is now applied and updated in draft mode as in previous versions (≤ Odoo 17).

We’ve tested it thoroughly — the deductible and non-deductible VAT amounts are recalculated correctly when editing draft invoices.

javierizaca reviewed (changes requested):
Hi @BetoRaja 

I've tested the new changes and found some unexpected behavior when resetting the invoice to draft:

1. The prorated breakdown is correct in "Draft" status.

2. If I move the invoice to "Posted", the amount remains correct.

3. When I reset the invoice to Draft, the amount of deductible prorated VAT increases to match 100% of the tax in the journal entries (see attached screenshot), while still keeping the non-deductible portion recorded as well.

<img width="1241" height="828" alt="image" src="https://github.com/user-attachments/assets/eabec676-9208-473c-a460-e6c411ad12e3" />

At this point, if I validate the invoice again, the amount due includes:

- Product
- 100% of the tax amount (VAT in this case)
- The portion of the tax not subject to prorated deduction (VAT in this case)

If I change the invoice date and there is a different prorated percentage configured for that date, saving the record triggers a full reset — everything is recalculated and corrected properly. However, once I validate the invoice and then set it back to draft, the issue appears again.

Thanks

ValentinVinagre commented on l10n_es_vat_prorate/models/account_move.py:
```diff
@@ -3,116 +3,176 @@
+        moves._apply_vat_prorate()
+        return moves
+
+    def write(self, vals):
```
Be careful with canceling the call to the super.
On the other hand, be careful with the application of the prorated rate, as there will be cases that are not covered by this method.

ValentinVinagre replied:
@BetoRaja feedback?

pedrobaeza committed:
[IMP+FIX+I18N] l10n_es_vat_prorate: Feedback from v16 backport

- account.move~with_special_vat_prorate can't depend on company prorate
  modifications, or this will trigger the line check computation,
  marking/unmarking all the existing invoices.
- Remove double assignation vat_prorate = False.
- Terms homogenized.
- Make prorate tree editable again.
- Better text for invoice line column.
- Translations.

pedrobaeza commented:
A tener en cuenta #4288, especialmente https://github.com/OCA/l10n-spain/pull/4288/files#diff-b98482b0eba53728bc7ae07431251e2bd1d61f5fc7001e15a80316693efe1b74R21

BetoRaja commented:
> A tener en cuenta #4288, especialmente https://github.com/OCA/l10n-spain/pull/4288/files#diff-b98482b0eba53728bc7ae07431251e2bd1d61f5fc7001e15a80316693efe1b74R21

@pedrobaeza Fixes https://github.com/OCA/l10n-spain/pull/4288 y http://github.com/OCA/l10n-spain/pull/4279 revisados y tenidos en cuenta para nuestra migración,Muchas gracias.

Commit:
[IMP]l10n_es_vat_prorate: pre-commit auto fixes

Commit:
[MIG] l10n_es_vat_prorate: Migration to 18.0

Jaimermaccione reviewed (approved):
@BetoRaja

Tested the case reported in the latest revision by @Javierizaca, specifically the behavior when resetting the invoice to *Draft* after being *Posted*.

All functionality works as expected:
- The prorated VAT calculation behaves correctly.
- Reverting to draft and revalidating the invoice maintains the correct deductible and non-deductible amounts.
- Changing the invoice date with different prorated percentages triggers accurate recalculation.

Tested on Runboat. Everything looks good.

---

#### LGTM!  
Thanks!

javierizaca reviewed (approved):
All functionality works as expected in the last version tested

Jesarregui reviewed (approved):
Functional Review: LGTM

ValentinVinagre reviewed (approved):
LGTM 👍🏻

ValentinVinagre commented:
and review it again 😄

ValentinVinagre reviewed (changes requested):
Tests

Anxo82 committed:
l10n_es_vat_prorate: Fix error test (Status Update 06/10/2025)
- Fixed test error: "Any journal item on a receivable account must have a due date and vice versa." https://github.com/OCA/l10n-spain/actions/runs/17910601677/job/50921072424?pr=4251
- Added condition so it only applies to vendor bills.
- New error detected when creating refund invoices: "The entry is not balanced."
- New error found in tax groups: amounts are not being distributed correctly.

EmilioPascual commented:
MT-11110 @moduon

EmilioPascual commented:
@BetoRaja parece que hay algún error con los test de la prorrata en el SII ¿Puedes revisarlo por favor? Gracias

HaraldPanten commented:
@EmilioPascual la persona encargada está de vacaciones unos días, pero vamos a retomarlo en breve.

Shide commented:
@BetoRaja @Anxo82 siguen fallando los tests con el SII, podéis echarle un ojo? Gracias!

Anxo82 commented:
@Shide si, estamos en ello, esperamos solucionarlo lo antes posible, perdonad la demora. Gracias

Anxo82 commented:
**Status Update (06/10/2025)**

- [x] Fixed test error: "Any journal item on a receivable account must have a due date and vice versa." (https://github.com/OCA/l10n-spain/actions/runs/17910601677/job/50921072424?pr=4251)
- [x] Added condition so it only applies to vendor bills. 
- [x] New error found in tax groups: amounts are not being distributed correctly.
<img width="2396" height="832" alt="image" src="https://github.com/user-attachments/assets/f2006e3a-c497-4782-ba7c-28ad90a49676" />

- [ ] New error detected when creating refund invoices: "The entry is not balanced."
- [ ] 
<img width="2346" height="420" alt="image" src="https://github.com/user-attachments/assets/f779ef2d-6ca7-4079-bda1-8d413a39e4c1" />


We are working on fixing these new issues. Thanks.

Anxo82 committed:
[WIP] vat_prorate: Continue work on refund/prorate logic (errors pending)

Anxo82 commented:
**Status Update (21/10/2025)**

I’m still working on the migration of the VAT prorate module.

- [x] Refund invoices (out_refund) are still failing, so further debugging is ongoing.
- [ ] Pending adaptation of the test to the new code
- [ ] I noticed that lines with with_special_vat_prorate=True are not correctly setting with_vat_prorate=True in AccountMoveLine. This means that VAT prorate is not applied for special proration moves. This is a pending fix and should be addressed in a future update.

Any feedback or contributions are very welcome. Thanks

Anxo82 committed:
[WIP] l10n_es_vat_prorate: Continue work

Jaimermaccione reviewed (changes requested):
@Anxo82 Functional Review – Tested on Runboat Instance

A few issues were detected during the review:

---

#### 1. Missing visibility of `special_vat_prorate_default` in the company configuration

In the company form, under the **"VAT Prorate"** tab, the field `special_vat_prorate_default` is not shown in the list view.  
This makes it difficult to configure whether invoice lines should be automatically marked when a special prorate applies.
<img width="1635" height="518" alt="image" src="https://github.com/user-attachments/assets/ab6c4f0f-0693-4591-a078-cff33f226aac" />
<img width="1263" height="490" alt="image" src="https://github.com/user-attachments/assets/ab3dbd9f-5848-42c3-bdfc-900f5fb2b478" />





---

#### 2. The `with_vat_prorate` field on invoice lines has no effect on accounting

Invoice lines can be automatically marked with the `with_vat_prorate` field **when a special prorate is active and the field `special_vat_prorate_default` is enabled** in the company prorate configuration.

However, even when lines are marked, this field currently has **no impact on accounting**.  
Journal entries and tax amounts remain the same.  
In v17, when properly configured, the with_vat_prorate field does affect tax calculation: the prorated amount is recalculated depending on whether the invoice line is marked or not.

---

#### 3. Prorate selection is based on the accounting date

The prorate percentage applied to a purchase invoice is determined by the invoice’s **accounting date** (technical field: `date`).  
This works correctly. Changing the date triggers a re-evaluation of the applicable prorate.

**Add to README:**

> The prorate value applied to each invoice is determined by the accounting date (`date`).  
> Ensure the company has valid prorate records for the corresponding periods.

---

#### General Notes

- The module works correctly when a valid prorate is configured for the company.  
- Recalculation based on date changes in draft invoices functions as expected.

Tisho99 commented on l10n_es_vat_prorate/models/account_move.py:
```diff
@@ -0,0 +1,232 @@
+    def _apply_vat_prorate(self):
+        if self.env.context.get("skip_vat_prorate"):
+            return
+        for move in self:
```
Aqui creo que hay que omitir el move si ves que no tiene el campo with_vat_prorate establecido , de forma similar a como se hace en https://github.com/OCA/l10n-spain/blob/17.0/l10n_es_vat_prorate/models/account_move.py#L79

Con la prorrata general, se aplica a todos los moves, pero con la especial no y hay que descartar por eso algunas lineas

HaraldPanten commented:
Cerramos PR en favor de https://github.com/OCA/l10n-spain/pull/4442 para seguir avanzando en la migración de la prorrata.
